numpyro/distributions/distribution.py

```diff
@@ -725,9 +725,23 @@ def __init__(self, base_distribution, transforms, validate_args=None):
         # to pay attention to any inference function that inspects
         # transformed distribution's shape.
         shape = base_distribution.batch_shape + base_distribution.event_shape
-        event_dim = max([len(base_distribution.event_shape)] + [t.event_dim for t in transforms])
-        batch_shape = shape[:len(shape) - event_dim]
-        event_shape = shape[len(shape) - event_dim:]
+        base_ndim = len(shape)
+        transform = ComposeTransform(self.transforms)
+        transform_input_event_dim = transform.input_event_dim
+        if base_ndim < transform_input_event_dim:
+            raise ValueError("Base distribution needs to have shape with size at least {}, but got {}."
+                             .format(transform_input_event_dim, base_ndim))
+        event_dim = transform.output_event_dim + max(self.base_dist.event_dim - transform_input_event_dim, 0)
+        # See the above note. Currently, there is no way to interpret the shape of output after
+        # transforming. To solve this issue, we need something like Bijector.forward_event_shape
+        # as in TFP. For now, we will prepend singleton dimensions to compromise, so that
+        # event_dim, len(batch_shape) are still correct.
+        if event_dim <= base_ndim:
+            batch_shape = shape[:base_ndim - event_dim]
+            event_shape = shape[base_ndim - event_dim:]
+        else:
+            event_shape = (1,) * (event_dim - base_ndim) + shape
```

Comment:
Is there a reason to prepend these singleton dims instead of suffixing these?

Reply:
Nope, I think we can also add singleton dims to the last. Either way will not be correct, but we only need to maintain the correct event_dim here.

Reply:
Hmm..I was thinking that the latter might be more accurate if we compare at each dimension, but it seems like both might be equally wrong. If for example, we have a LowerCholesky constraint on a univariate normal with `batch_shape=(3,)`, then actual event shape = `(2, 2` (assuming we include the diagonal), but this will give us `(1, 3)` which is completely wrong. Is having `shape` useful in other contexts?

Reply:
I found only three places using `.event_shape` or `.shape()`: in [funsor](https://github.com/pyro-ppl/numpyro/blob/master/numpyro/contrib/funsor/enum_messenger.py#L529), [autoguide](https://github.com/pyro-ppl/numpyro/blob/master/numpyro/infer/autoguide.py#L420), and [scan](https://github.com/pyro-ppl/numpyro/blob/master/numpyro/contrib/control_flow/scan.py#L60). All of them only use `event_shape` to calculate `event_dim`. In my opinion, only `batch_shape` and `event_dim` are useful in inference algorithms. :)

Reply:
I agree with that assessment. What I was trying to get at is should we make it more obvious that we don't know what the event shape is, by say having `(-1,) * event_dim` instead of having something that sort of looks correct but can be buggy? That will broadcast fine to a single dim and otherwise raise a loud error.

Reply:
Thanks, your suggestion looks good to me!